src/Compilers/Core/Portable/SourceGeneration/Nodes/NodeStateTable.cs

```diff
@@ -374,17 +373,26 @@ public bool TryModifyEntries(ImmutableArray<T> outputs, IEqualityComparer<T> com
             public void AddEntry(T value, EntryState state, TimeSpan elapsedTime, ImmutableArray<(IncrementalGeneratorRunStep InputStep, int OutputIndex)> stepInputs, EntryState overallInputState)
             {
                 _states.Add(new TableEntry(OneOrMany.Create(value), state));
+                _insertedCount += state == EntryState.Added ? 1 : 0;
                 RecordStepInfoForLastEntry(elapsedTime, stepInputs, overallInputState);
             }
 
             public TableEntry AddEntries(ImmutableArray<T> values, EntryState state, TimeSpan elapsedTime, ImmutableArray<(IncrementalGeneratorRunStep InputStep, int OutputIndex)> stepInputs, EntryState overallInputState)
             {
                 var tableEntry = new TableEntry(OneOrMany.Create(values), state);
                 _states.Add(tableEntry);
+                _insertedCount += state == EntryState.Added ? 1 : 0;
```

Comment:
Won't the problem occur when an entry is inserted with state other than `Added`?

Reply:
Sort of. AddEntries is used both as `AppendEntries` and also `ReplaceEntries`. When called with `EntryState.Added` it means add a new entry, so we need to track the insertion. Everything else that calls it with a different entry state is replacing the exisitng entry with something else.

It's not a good API (read: it is terribad). I'm planning on refactoring the StateTable API anyway, so it always maintains consistency rather than relying on the callers to it to use it correctly, but that's a bigger change and I want to make the smallest one possible here as a targeted fix.